File: python/baiduads-sdk-samples/baiduads/samples/search/create_search_ad.py
"""
创建搜索广告示例
"""
import logging

from baiduads.adgroup.api import AdgroupService
from baiduads.adgroup.model import AddAdgroupRequestWrapper, ApiAdgroupAddRequest, ApiAdgroupType
from baiduads.campaign.api import CampaignService
from baiduads.campaign.model import AddCampaignRequestWrapper, ApiCampaignAddRequest, ApiCampaignType, ApiScheduleType, \
    SchedulePriceFactor, RegionPriceFactor
from baiduads.common.model import ApiRequestHeader
from baiduads.creative.api import CreativeService
from baiduads.creative.model import AddCreativeRequestWrapper, ApiCreativeAddRequest, ApiCreativeType
from baiduads.keyword.api import KeywordService
from baiduads.keyword.model import AddWordRequestWrapper, ApiKeywordAddRequest, ApiKeywordType

logger = logging.getLogger("CreateSearchAd")
logger.setLevel(logging.INFO)
# set STDOUT
logger.addHandler(logging.StreamHandler())


class CreateSearchAd:
    header = ApiRequestHeader(
        token="xxxxx",
        username="xxxxx",
        password="xxxxx",
        _spec_property_naming=True
    )

    @classmethod
    def add_campaign(cls):
        campaign_service = CampaignService()
        # 构造参数
        schedule = ApiScheduleType(
            startHour=10,
            endHour=12,
            weekDay=6,
            _spec_property_naming=True
        )

        region_price_factor = RegionPriceFactor(
            regionId=1000,
            priceFactor=8.8,
            _spec_property_naming=True
        )

        schedule_price_factor = SchedulePriceFactor(
            timeId=123,
            priceFactor=6.6,
            _spec_property_naming=True
        )

        campaign_type = ApiCampaignType(
            campaignName="xx计划名称",
            budget=50.0,
            regionTarget=[1000, 4085],
            negativeWords=["短语否词1", "短语否词2"],
            exactNegativeWords=["精确否词1", "精确否词2"],
            schedule=[schedule],
            pause=True,
            priceRatio=2.0,
            bidPrefer=1,
            adType=0,
            businessPointId=123456789,
            regionPriceFactor=[region_price_factor],
            schedulePriceFactors=[schedule_price_factor],
            marketingTargetId=0,
            _spec_property_naming=True
        )
        api_campaign_add_request = ApiCampaignAddRequest(
            adType=0,
            campaignTypes=[campaign_type],
            _spec_property_naming=True
        )
        add_campaign_request_wrapper = AddCampaignRequestWrapper(
            header=cls.header,  # 设置请求header
            body=api_campaign_add_request,  # 设置请求body
            _spec_property_naming=True
        )
        # 调用API
        add_campaign_response_wrapper = campaign_service.add_campaign(add_campaign_request_wrapper)
        # 判断请求是否成功
        if add_campaign_response_wrapper.header.status != 0:
            logger.error("add_campaign fail. errors:%s", add_campaign_response_wrapper.header)
            raise RuntimeError("add_campaign fail.")
        return add_campaign_response_wrapper.body.data

    @classmethod
    def add_adgroup(cls, campaign_id):
        adgroup_service = AdgroupService()
        # 构造参数
        api_adgroup_type = ApiAdgroupType(
            campaignId=campaign_id,
            adgroupName="xx单元名称",
            maxPrice=6.8,
            pause=True,
            negativeWords=["短语否词1", "短语否词2"],
            exactNegativeWords=["精确否词1", "精确否词2"],
            priceRatio=5.0,
            segmentRecommendStatus=0,
            _spec_property_naming=True
        )

        api_adgroup_add_request = ApiAdgroupAddRequest(
            adgroupTypes=[api_adgroup_type],
            _spec_property_naming=True
        )
        add_adgroup_request_wrapper = AddAdgroupRequestWrapper(
            header=cls.header,
            body=api_adgroup_add_request,
            _spec_property_naming=True
        )
        add_adgroup_response_wrapper = adgroup_service.add_adgroup(add_adgroup_request_wrapper)
        if add_adgroup_response_wrapper.header.status != 0:
            logger.error("add_adgroup fail. error:%s", add_adgroup_response_wrapper.header)
            raise RuntimeError("add_adgroup fail.")
        return add_adgroup_response_wrapper.body.data

    @classmethod
    def add_creative(cls, campaign_id, adgroup_id):
        creative_service = CreativeService()
        # 构造参数
        api_creative_type = ApiCreativeType(
            campaignId=campaign_id,
            adgroupId=adgroup_id,
            title="xx创意标题",
            description1="针对数据不完整情况,我们有成熟的解决方案,让{数据}完整恢复",
            description2="恢复成功才收费{中毒}数据恢复值得信赖",
            mobileDestinationUrl="https://aisite.wejianzhan.com/site/abcd.com/46aa6sssd1b-d206-4dfc-999a-7b8b487aab5d",
            mobileDisplayUrl="abcd.com",
            pcDestinationUrl="https://aisite.wejianzhan.com/site/abcd.com/46aa6sssd1b-d206-4dfc-999a-7b8b487aab5d",
            pcDisplayUrl="abcd.com",
            tabs=[1]
        )

        api_creative_request = ApiCreativeAddRequest(
            creativeTypes=[api_creative_type],
            _spec_property_naming=True
        )
        add_creative_request_wrapper = AddCreativeRequestWrapper(
            header=cls.header,  # 设置请求header
            body=api_creative_request,  # 设置请求body
            _spec_property_naming=True
        )
        # 调用API
        add_creative_response_wrapper = creative_service.add_creative(add_creative_request_wrapper)
        # 判断请求是否成功
        if add_creative_response_wrapper.header.status != 0:
            logger.error("add_creative fail. errors:%s", add_creative_response_wrapper.header)
            raise RuntimeError("add_creative fail.")
        return add_creative_response_wrapper.body.data

    @classmethod
    def add_word(cls, adgroup_id):
        keyword_service = KeywordService()
        # 构造参数
        api_keyword_type = ApiKeywordType(
            adgroupId=adgroup_id,
            keyword="xx关键词名称",
            price=9.8,
            matchType=2,
            phraseType=3,
            pause=True,
            tabs=[1, 31],
            pcDestinationUrl="https://aisite.wejianzhan.com/site/abcd.com/46aasd1b-d206-4dfc-999a-7b8b487aab",
            mobileDestinationUrl="https://aisite.wejianzhan.com/site/abcd.com/46aasd1b-d206-4dfc-999a-7b8b487aab",
            _spec_property_naming=True
        )
        api_keyword_add_request = ApiKeywordAddRequest(
            keywordTypes=[api_keyword_type],
            _spec_property_naming=True
        )
        add_word_request_wrapper = AddWordRequestWrapper(
            header=cls.header,  # 设置请求header
            body=api_keyword_add_request,  # 设置请求body
            _spec_property_naming=True
        )
        # 调用API
        add_word_response_wrapper = keyword_service.add_word(add_word_request_wrapper)
        # 判断请求是否成功
        if add_word_response_wrapper.header.status != 0:
            logger.error("add_word fail. errors:%s", add_word_response_wrapper.header)
            raise RuntimeError("add_word fail.")
        return add_word_response_wrapper.body.data


def main():
    # 第一步：新建计划
    api_campaign_types = CreateSearchAd.add_campaign()
    logger.info("api_campaign_types:%s", api_campaign_types)
    # 获取campaignId
    campaign_id = api_campaign_types[0].campaign_id

    # 第二步：新建单元
    api_adgroup_types = CreateSearchAd.add_adgroup(campaign_id)
    logger.info("api_adgroup_types:%s", api_adgroup_types)
    # 获取adgroup_id
    adgroup_id = api_adgroup_types[0].adgroup_id

    # 第三步：在单元下添加创意
    api_creative_types = CreateSearchAd.add_creative(campaign_id, adgroup_id)
    logger.info("api_creative_types:%s", api_creative_types)

    # 第四步：在单元下添加关键词
    api_keyword_types = CreateSearchAd.add_word(adgroup_id)
    logger.info("api_keyword_types:%s", api_keyword_types)


if __name__ == '__main__':
    main()

File: java/baiduads-sdk-samples/src/main/java/com/baidu/dev2/api/samples/search/CreateSearchAd.java
package com.baidu.dev2.api.samples.search;

import com.baidu.dev2.api.sdk.adgroup.api.AdgroupService;
import com.baidu.dev2.api.sdk.adgroup.model.AddAdgroupRequestWrapper;
import com.baidu.dev2.api.sdk.adgroup.model.AddAdgroupResponseWrapper;
import com.baidu.dev2.api.sdk.adgroup.model.ApiAdgroupAddRequest;
import com.baidu.dev2.api.sdk.adgroup.model.ApiAdgroupType;
import com.baidu.dev2.api.sdk.campaign.api.CampaignService;
import com.baidu.dev2.api.sdk.campaign.model.AddCampaignRequestWrapper;
import com.baidu.dev2.api.sdk.campaign.model.AddCampaignResponseWrapper;
import com.baidu.dev2.api.sdk.campaign.model.ApiCampaignAddRequest;
import com.baidu.dev2.api.sdk.campaign.model.ApiCampaignType;
import com.baidu.dev2.api.sdk.campaign.model.ApiScheduleType;
import com.baidu.dev2.api.sdk.campaign.model.RegionPriceFactor;
import com.baidu.dev2.api.sdk.campaign.model.SchedulePriceFactor;
import com.baidu.dev2.api.sdk.common.ApiRequestHeader;
import com.baidu.dev2.api.sdk.creative.api.CreativeService;
import com.baidu.dev2.api.sdk.creative.model.AddCreativeRequestWrapper;
import com.baidu.dev2.api.sdk.creative.model.AddCreativeResponseWrapper;
import com.baidu.dev2.api.sdk.creative.model.ApiCreativeAddRequest;
import com.baidu.dev2.api.sdk.creative.model.ApiCreativeType;
import com.baidu.dev2.api.sdk.invoke.ApiException;
import com.baidu.dev2.api.sdk.keyword.api.KeywordService;
import com.baidu.dev2.api.sdk.keyword.model.AddWordRequestWrapper;
import com.baidu.dev2.api.sdk.keyword.model.AddWordResponseWrapper;
import com.baidu.dev2.api.sdk.keyword.model.ApiKeywordAddRequest;
import com.baidu.dev2.api.sdk.keyword.model.ApiKeywordType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;

/**
 * 创建搜索广告示例
 */
public class CreateSearchAd {

    private static final Logger LOGGER = LoggerFactory.getLogger(CreateSearchAd.class);

    private static ApiRequestHeader header;

    static {
        header = new ApiRequestHeader();
        header.setPassword("xxxxx");
        header.setUsername("xxxxx");
        header.setToken("xxxxx");
    }

    /**
     * 新建计划
     *
     * @return
     * @throws ApiException
     */
    private static List<ApiCampaignType> addCampaign() throws ApiException {
        CampaignService campaignService = new CampaignService();
        AddCampaignRequestWrapper addCampaignRequestWrapper = new AddCampaignRequestWrapper();
        // 设置请求header
        addCampaignRequestWrapper.setHeader(header);
        //  构造请求参数
        ApiCampaignAddRequest apiCampaignAddRequest = new ApiCampaignAddRequest();
        apiCampaignAddRequest.setAdType(0);
        List<ApiCampaignType> campaignTypes = new ArrayList<>();
        ApiCampaignType apiCampaignType = new ApiCampaignType();
        apiCampaignType.setCampaignName("xx计划名称");
        apiCampaignType.budget(50.0);
        // 设置计划推广地域
        List<Integer> regionTarget = new ArrayList<>();
        regionTarget.add(1000);
        regionTarget.add(4085);
        apiCampaignType.regionTarget(regionTarget);
        List<String> negativeWords = new ArrayList<>();
        negativeWords.add("短语否词1");
        negativeWords.add("短语否词2");
        apiCampaignType.setNegativeWords(negativeWords);
        List<String> exactNegativeWords = new ArrayList<>();
        exactNegativeWords.add("精确否词1");
        exactNegativeWords.add("精确否词2");
        apiCampaignType.setExactNegativeWords(exactNegativeWords);
        // 设置计划推广暂停时段
        List<ApiScheduleType> schedule = new ArrayList<>();
        ApiScheduleType apiScheduleType = new ApiScheduleType();
        apiScheduleType.setStartHour(10L);
        apiScheduleType.setEndHour(12L);
        apiScheduleType.setWeekDay(6L);
        schedule.add(apiScheduleType);
        apiCampaignType.setSchedule(schedule);

        apiCampaignType.setPause(true);
        apiCampaignType.setPriceRatio(2.0);
        apiCampaignType.bidPrefer(1);
        apiCampaignType.setAdType(0);
        apiCampaignType.setBusinessPointId(123456789L);
        // 分地域出价系数
        List<RegionPriceFactor> regionPriceFactors = new ArrayList<>();
        RegionPriceFactor regionPriceFactor = new RegionPriceFactor();
        regionPriceFactor.setRegionId(1000);
        regionPriceFactor.setPriceFactor(8.8);
        regionPriceFactors.add(regionPriceFactor);
        apiCampaignType.setRegionPriceFactor(regionPriceFactors);
        // 分时段出价系数
        List<SchedulePriceFactor> schedulePriceFactors = new ArrayList<>();
        SchedulePriceFactor schedulePriceFactor = new SchedulePriceFactor();
        schedulePriceFactor.setTimeId(123);
        schedulePriceFactor.setPriceFactor(6.6);
        schedulePriceFactors.add(schedulePriceFactor);
        apiCampaignType.setSchedulePriceFactors(schedulePriceFactors);

        apiCampaignType.setMarketingTargetId(0);
        campaignTypes.add(apiCampaignType);
        apiCampaignAddRequest.setCampaignTypes(campaignTypes);
        // 设置请求body
        addCampaignRequestWrapper.setBody(apiCampaignAddRequest);
        // 调用API
        AddCampaignResponseWrapper addCampaignResponseWrapper =
                campaignService.addCampaign(addCampaignRequestWrapper);
        // 判断请求是否成功
        if (addCampaignResponseWrapper.getHeader().getStatus() != 0) {
            LOGGER.error("addCampaign fail. errors:" + addCampaignResponseWrapper.getHeader());
            throw new RuntimeException("addCampaign fail.");
        }
        return addCampaignResponseWrapper.getBody().getData();
    }

    /**
     * 在新建计划下新建单元
     *
     * @param campaignId
     * @return
     * @throws ApiException
     */
    public static List<ApiAdgroupType> addAdgroup(Long campaignId) throws ApiException {
        AdgroupService adgroupService = new AdgroupService();
        AddAdgroupRequestWrapper addAdgroupRequestWrapper = new AddAdgroupRequestWrapper();
        // 设置请求header
        addAdgroupRequestWrapper.setHeader(header);
        //  构造请求参数
        ApiAdgroupAddRequest apiAdgroupAddRequest = new ApiAdgroupAddRequest();
        List<ApiAdgroupType> adgroupTypes = new ArrayList<>();
        ApiAdgroupType apiAdgroupType = new ApiAdgroupType();
        apiAdgroupType.setCampaignId(campaignId);
        apiAdgroupType.setAdgroupName("xx单元名称");
        apiAdgroupType.setMaxPrice(6.8);
        apiAdgroupType.setPause(true);
        List<String> negativeWords = new ArrayList<>();
        negativeWords.add("短语否词1");
        negativeWords.add("短语否词2");
        apiAdgroupType.setNegativeWords(negativeWords);
        List<String> exactNegativeWords = new ArrayList<>();
        exactNegativeWords.add("精确否词1");
        exactNegativeWords.add("精确否词2");
        apiAdgroupType.setExactNegativeWords(exactNegativeWords);
        apiAdgroupType.setPriceRatio(5.0);
        apiAdgroupType.setSegmentRecommendStatus(0);
        adgroupTypes.add(apiAdgroupType);
        apiAdgroupAddRequest.setAdgroupTypes(adgroupTypes);
        // 设置请求body
        addAdgroupRequestWrapper.setBody(apiAdgroupAddRequest);
        // 调用API
        AddAdgroupResponseWrapper addAdgroupResponseWrapper = adgroupService.addAdgroup(addAdgroupRequestWrapper);

        // 判断请求是否成功
        if (addAdgroupResponseWrapper.getHeader().getStatus() != 0) {
            LOGGER.error("addAdgroup fail. errors:" + addAdgroupResponseWrapper.getHeader());
            throw new RuntimeException("addAdgroup fail.");
        }
        return addAdgroupResponseWrapper.getBody().getData();
    }

    /**
     * 在单元下添加创意
     *
     * @param campaignId
     * @param adgroupId
     * @return
     * @throws ApiException
     */
    public static List<ApiCreativeType> addCreative(Long campaignId, Long adgroupId) throws ApiException {
        CreativeService creativeService = new CreativeService();
        AddCreativeRequestWrapper addCreativeRequestWrapper = new AddCreativeRequestWrapper();
        // 设置请求header
        addCreativeRequestWrapper.setHeader(header);
        // 构造请求参数
        ApiCreativeAddRequest apiCreativeAddRequest = new ApiCreativeAddRequest();
        List<ApiCreativeType> creativeTypes = new ArrayList<>();
        ApiCreativeType apiCreativeType = new ApiCreativeType();
        apiCreativeType.setCampaignId(campaignId);
        apiCreativeType.setAdgroupId(adgroupId);
        apiCreativeType.setTitle("xx创意标题");
        apiCreativeType.setDescription1("针对数据不完整情况,我们有成熟的解决方案,让{数据}完整恢复");
        apiCreativeType.setDescription2("恢复成功才收费{中毒}数据恢复值得信赖");
        String moDestinationUrl = "https://aisite.wejianzhan.com/site/abcd.com/46aa6sssd1b-d206-4dfc-999a-7b8b487aab5d";
        apiCreativeType.setMobileDestinationUrl(moDestinationUrl);
        apiCreativeType.setMobileDisplayUrl("abcd.com");
        String pcDestinationUrl = "https://aisite.wejianzhan.com/site/abcd.com/46aa6sssd1b-d206-4dfc-999a-7b8b487aab5d";
        apiCreativeType.setPcDestinationUrl(pcDestinationUrl);
        apiCreativeType.setPcDisplayUrl("abcd.com");
        List<Integer> tabs = new ArrayList<>();
        tabs.add(1);
        apiCreativeType.setTabs(tabs);
        creativeTypes.add(apiCreativeType);
        apiCreativeAddRequest.setCreativeTypes(creativeTypes);
        // 设置请求body
        addCreativeRequestWrapper.setBody(apiCreativeAddRequest);
        // 调用API
        AddCreativeResponseWrapper addCreativeResponseWrapper =
                creativeService.addCreative(addCreativeRequestWrapper);

        // 判断请求是否成功
        if (addCreativeResponseWrapper.getHeader().getStatus() != 0) {
            LOGGER.error("addCreative fail. errors:" + addCreativeResponseWrapper.getHeader());
            throw new RuntimeException("addCreative fail.");
        }
        return addCreativeResponseWrapper.getBody().getData();
    }

    /**
     * 在单元下添加关键词
     *
     * @param adgroupId
     * @return
     * @throws ApiException
     */
    public static List<ApiKeywordType> addWord(Long adgroupId) throws ApiException {
        KeywordService keywordService = new KeywordService();
        AddWordRequestWrapper addWordRequestWrapper = new AddWordRequestWrapper();
        // 设置请求header
        addWordRequestWrapper.setHeader(header);
        // 构造请求参数
        ApiKeywordAddRequest apiKeywordAddRequest = new ApiKeywordAddRequest();
        List<ApiKeywordType> keywordTypes = new ArrayList<>();
        ApiKeywordType apiKeywordType = new ApiKeywordType();
        apiKeywordType.setAdgroupId(adgroupId);
        apiKeywordType.setKeyword("xx关键词名称");
        apiKeywordType.setPrice(9.8);
        apiKeywordType.setMatchType(2);
        apiKeywordType.setPhraseType(3);
        apiKeywordType.setPause(true);
        List<Integer> tabs = new ArrayList<>();
        tabs.add(1);
        tabs.add(31);
        apiKeywordType.setTabs(tabs);
        String pcDestinationUrl = "https://aisite.wejianzhan.com/site/abcd.com/46aa6sssd1b-d206-4dfc-999a-7b8b487aab5d";
        apiKeywordType.setPcDestinationUrl(pcDestinationUrl);
        String moDestinationUrl = "https://aisite.wejianzhan.com/site/abcd.com/46aa6sssd1b-d206-4dfc-999a-7b8b487aab5d";
        apiKeywordType.setMobileDestinationUrl(moDestinationUrl);
        keywordTypes.add(apiKeywordType);
        apiKeywordAddRequest.setKeywordTypes(keywordTypes);
        // 设置请求body
        addWordRequestWrapper.setBody(apiKeywordAddRequest);
        // 调用API
        AddWordResponseWrapper addWordResponseWrapper = keywordService.addWord(addWordRequestWrapper);

        // 判断请求是否成功
        if (addWordResponseWrapper.getHeader().getStatus() != 0) {
            LOGGER.error("addWord fail. errors:" + addWordResponseWrapper.getHeader());
            throw new RuntimeException("addWord fail.");
        }
        return addWordResponseWrapper.getBody().getData();
    }

    public static void main(String[] args) {
        try {
            // 第一步：创建计划
            List<ApiCampaignType> apiCampaignTypes = addCampaign();
            LOGGER.info("apiCampaignTypes:{}", apiCampaignTypes);
            // 获取计划id
            Long campaignId = apiCampaignTypes.get(0).getCampaignId();

            // 第二步：在计划下新建单元
            List<ApiAdgroupType> apiAdgroupTypes = addAdgroup(campaignId);
            LOGGER.info("apiAdgroupTypes:{}", apiAdgroupTypes);
            // 获取单元id
            Long adgroupId = apiAdgroupTypes.get(0).getAdgroupId();

            // 第三步：在单元下添加创意
            List<ApiCreativeType> apiCreativeTypes = addCreative(campaignId, adgroupId);
            LOGGER.info("apiCreativeTypes:{}", apiCreativeTypes);
            // 第四步：在单元下添加关键词
            List<ApiKeywordType> apiKeywordTypes = addWord(adgroupId);
            LOGGER.info("apiKeywordTypes:{}", apiKeywordTypes);
        } catch (Exception e) {
            LOGGER.error("CreateSearchAd fail.", e);
        }
    }
}
